tests/unit/s2n_async_pkcs11_sign_test.c

```diff
@@ -0,0 +1,574 @@
+        mechanism.pParameter= NULL;
+        mechanism.ulParameterLen = 0 ;
+    }
+
+    POSIX_GUARD(function_list->C_SignInit(session,
+                                           &mechanism,
+                                           key));
+
+
+    POSIX_GUARD(function_list->C_Sign(session,
+                                       temp_digest,
+                                       temp_digest_len,
+                                       NULL,
+                                       (CK_ULONG_PTR)sig_len));
+
+    /* We need an extra 8 bytes to store the ASN.1 encoding if the key is ECDSA. */
+    uint8_t * sig = malloc(*sig_len + 8);
+    POSIX_GUARD_PTR(sig);
+    POSIX_GUARD(function_list->C_Sign(session,
+                                       temp_digest,
+                                       temp_digest_len,
+                                       sig,
+                                       (CK_ULONG_PTR)sig_len));
+    if(keytype == CKK_ECDSA)
+    {
+        /* Convert from a PKCS #11 encoding to an ASN.1 encoding, sig_len will be updated to account for the extra bytes in the conversion. */
+        pkcs11_sig_to_asn1_sig(sig, sig_len);
+    }
+
+   *sig_buf = sig;
+    pkcs11_cleanup(session);
+
+    return 0;
+}
+
+void * pkey_task(void * params)
+{
+    struct task_params * info = (struct task_params *) params;
+
+    struct s2n_connection *conn = info->conn;
+    struct s2n_async_pkey_op *op = info->op;
+
+    uint32_t input_len;
+    s2n_async_pkey_op_get_input_size(op, &input_len);
+    uint8_t * input = malloc(input_len);
+
+    s2n_async_pkey_op_get_input(op, input,input_len);
+
+    uint8_t * output = NULL;
+    uint32_t output_len = 0;
+
+    s2n_async_pkey_op_type type;
+    s2n_async_get_op_type(op, &type);
+
+    pthread_mutex_lock(&pkcs11_mutex);
+
+    CK_OBJECT_HANDLE handle;
+    CK_SESSION_HANDLE session;
+
+    struct s2n_cert_chain_and_key * cert_key = s2n_connection_get_selected_cert(conn);
+    char * label = s2n_cert_chain_and_key_get_ctx(cert_key);
+
+    pkcs11_setup(&session, &handle, label, strlen(label));
+    if(type == S2N_ASYNC_DECRYPT)
+    {
+        pkcs11_decrypt(session, handle, input, input_len, &output, &output_len);
+    } else {
+        pkcs11_sign(session, handle, input, input_len, &output, &output_len);
+    }
+    pthread_mutex_unlock(&pkcs11_mutex);
+
+    s2n_async_pkey_copy_output(op, output, output_len);
+    free(output);
+
+    s2n_async_pkey_op_apply(op, conn);
+    
+    free(params);
+    pthread_exit(NULL);
+}
+
+int async_pkey_callback(struct s2n_connection *conn, struct s2n_async_pkey_op *op)
+{
+    EXPECT_NOT_NULL(op);
+    pthread_t worker;
+    struct task_params * params = malloc(sizeof(struct task_params));
+    params->conn = conn; 
+    params->op = op; 
+
+    POSIX_GUARD(pthread_create(&worker, NULL, &pkey_task, params));
+
+    return S2N_SUCCESS;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+    EXPECT_SUCCESS(s2n_disable_tls13());
+
```

Comment:
Do not call this method. It's deprecated.

Reply:
Oops, thanks